### PR TITLE
Use Phylanx from Python

### PR DESCRIPTION
**Pull request by stevenrbrandt**
This pull request creates ...
(1) a python file, Phylanx2.py, which constructs an AST without using a parser.
(2) a function called util.append_operation() because, apparently, appending to the rest member of expressions does not work.
(3) a function util.stringify() which renders an AST in human-readable form.

### CODE REVIEW

File: src/ast/node.cpp [hunk 1]
```diff
@@ -376,6 +377,10 @@ namespace phylanx { namespace ast
     std::ostream& operator<<(std::ostream& out, operand const& op)
     {
         out << "operand";
+        int ind = op.index();
+        if(ind == 1) {
+          auto ident = phylanx::util::get<1>(op.get());
+        }
```
Comment (hkaiser) [inline src/ast/node.cpp]: Shouldn't this print something for `ind == 2` as well? Wouldn't it be better to use visit()?

File: src/util/serialization/ast.cpp [hunk 2]
```diff
@@ -142,6 +143,139 @@ namespace phylanx { namespace util
         detail::unserialize_helper(input, expr);
         return expr;
     }
-}}
 
+    // You cannot add to expression.rest unless you
+    // use this function. Note sure why.
+    void append_operation(ast::expression& ast,ast::operation const& o) {
+      ast.rest.push_back(o);
+    }
```
Comment (hkaiser) [inline src/util/serialization/ast.cpp]: `expression has a member `append()` just for this (see: https://github.com/STEllAR-GROUP/phylanx/blob/master/phylanx/ast/node.hpp#L472-L487)

**Comment by hkaiser** [comment]
Please use the `.clang-format` file in the root of this repository to format your code.

**Comment by hkaiser** [comment]
Also, could you please add a test verifying the `stringify` functionality?

File: src/ast/node.cpp (hunk 1 above)
Comment (stevenrbrandt) [reply src/ast/node.cpp]: Sorry, that was a mistake. I reverted it.

File: src/util/serialization/ast.cpp (hunk 2 above)
Comment (stevenrbrandt) [reply src/util/serialization/ast.cpp]: OK, I thought that was the built-in pybind11 command for lists. It didn't work for me. Calling it left the list empty.

Comment (hkaiser) [reply src/util/serialization/ast.cpp]: Ok, good point - I'll have a look and fix this. Thanks!

**Comment by stevenrbrandt** [comment]
What verison of clang do I need to use for the .clang-format? I get an error from both 3.8.1 and 3.9.0 about 
```
YAML:26:23: error: unknown key 'AlignEscapedNewlines'
AlignEscapedNewlines: Right
                      ^~~~~
```

**Comment by aserio** [comment]
@stevenrbrandt I got clang-format to work with Clang v5.0.0

File: tests/unit/python/ast/python_builds_ast.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
+import phylanx.ast as ast
```
Comment (hkaiser) [inline tests/unit/python/ast/python_builds_ast.py]: Please add copyright and license to this file

**Comment by stevenrbrandt** [comment]
Hartmut, I've added the copyrights.

**Comment by hkaiser** [comment]
@stevenrbrandt AppVeyor is still failing. I think you missed to explicitly export the stringify functions (add a `PHYLANX_EXPORT`) for them to be visible outside of the shared library.

**Comment by hkaiser** [comment]
@stevenrbrandt What was your rationale of adding another set of functions (`stringify_xxx`) instead of extending the existing `operator<<()` functionality to give you the textual representation of the expressions?

**Comment by stevenrbrandt** [comment]
You were already using that for a different representation, one that showed the tree structure more, and I didn't want to mess that up.

**Comment by hkaiser** [comment]
> You were already using that for a different representation, one that showed the tree structure more, and I didn't want to mess that up.

@stevenrbrandt This was just a preliminary implementation which has been updated since (see https://github.com/STEllAR-GROUP/phylanx/blob/master/src/ast/node.cpp#L423-L541). Could you please verify if these (new) streaming operators expose the functionality you need? Should we amend them in any way?

This should also bind those operators to the existing Python `__str__` functionality (see https://github.com/STEllAR-GROUP/phylanx/blob/master/python/src/phylanx.cpp#L113-L119).

...sorry for any possible misunderstandings, I should have made clear (at least in a comment or so) that the old streaming operators were just placeholders.

File: phylanx/util/serialization/ast.hpp [hunk 4]
```diff
@@ -33,6 +33,15 @@ namespace phylanx { namespace util
     PHYLANX_EXPORT std::vector<char> serialize(ast::expression const&);
     PHYLANX_EXPORT std::vector<char> serialize(ast::function_call const&);
 
+    PHYLANX_EXPORT void append_operation(ast::expression &,ast::operation const &);
+    PHYLANX_EXPORT std::string stringify_optoken(ast::optoken);
```
Comment (hkaiser) [inline phylanx/util/serialization/ast.hpp]: I think these declarations can be removed now.

File: py/Phylanx2.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
+# Copyright (c) 2017 Hartmut Kaiser
```
Comment (hkaiser) [inline py/Phylanx2.py]: Shouldn't this copyright be under your name?

File: tests/unit/python/ast/python_builds_ast.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
+# Copyright (c) 2017 Hartmut Kaiser
```
Comment (hkaiser) [inline tests/unit/python/ast/python_builds_ast.py]: Same here, please attribute the copyright to yourself.

**Comment by stevenrbrandt** [comment]
It won't pass the test I put in until the fix_lists PR is in. On the other hand, this PR will provide the test for fix_lists. Should I combine these two pull requests?

**Comment by hkaiser** [comment]
No need to combine things., Please fix the small issues I mentioned above and we'll merge this..

**Pull request merged**